### PR TITLE
add skip ci functionality

### LINKED ISSUE #28020: Support skipping Gitea Action workflows [ci skip] [no ci] ...
### Description

Hello, thanks everyone for the awesome Gitea Actions feature!

GitHub Actions (and many other CI tools, too) has a feature that allows the user to skip running CI when a particular keyword is part of the commit message.
GitHub documentation: https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

> You can skip workflow runs triggered by the `push` and `pull_request` events by including a command in your commit message:  
>    [skip ci]
>    [ci skip]
>    [no ci]
>    [skip actions]
>    [actions skip]

It seems that this is currently not supported by Gitea (can anyone else confirm this?).
Short term, it should be added to the list of Unsupported features: https://docs.gitea.com/next/usage/actions/comparison
Medium term we should of course implement this feature.

### Gitea Version

1.20.5

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

Screenshot 2023-11-13 at 16-56-33 blog

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Gitea Helm chart version 9.5.1: https://gitea.com/gitea/helm-chart/

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by denyskon**
Adds the possibility to skip workflow execution if the commit message contains a string like [skip ci] or similar.

The default strings are the same as on GitHub, users can also set custom ones in app.ini

Reference: https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

Close #28020

### CODE REVIEW

File: services/actions/notifier_helper.go [hunk 1]
```diff
@@ -144,6 +145,17 @@ func notify(ctx context.Context, input *notifyInput) error {
 		return fmt.Errorf("gitRepo.GetCommit: %w", err)
 	}
 
+	if input.Event == webhook_module.HookEventPush || input.Event == webhook_module.HookEventPullRequest {
+		// skip runs with skip ci prefix in commit message if the event is push or pull_request
+		// https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs
+		for _, prefix := range setting.Actions.SkipRunPrefix {
+			if strings.HasPrefix(commit.CommitMessage, prefix) {
```
Comment (ChristopherHX) [inline services/actions/notifier_helper.go]: Just a quick thought while reading prefix in this context.
_Isn't it contains on GitHub Actions? I'm used to place it anywhere in the CommitMessage on GitHub._

Comment (denyskon) [reply services/actions/notifier_helper.go]: Thanks, you're right. That's what happens when you implement a feature you never used 😆

**Comment by delvh** [comment]
I'll be probably be able to review on saturday.

File: services/actions/notifier_helper.go (hunk 1 above)
Comment (denyskon) [reply services/actions/notifier_helper.go]: Changed

**Review by lunny — APPROVED** [review]

File: services/actions/notifier_helper.go [hunk 2]
```diff
@@ -144,6 +145,17 @@ func notify(ctx context.Context, input *notifyInput) error {
 		return fmt.Errorf("gitRepo.GetCommit: %w", err)
 	}
 
+	if input.Event == webhook_module.HookEventPush || input.Event == webhook_module.HookEventPullRequest {
```
Comment (nanguanlin6) [inline services/actions/notifier_helper.go]: In Gitea, when a new commit is pushed to PR's head branch, the webhook event is called `HookEventPullRequestSync`, which should also trigger this `skip ci` functionality. See Github example in my test repo(https://github.com/lng2020/fork_repo/pull/1)

**Review by nanguanlin6 — CHANGES_REQUESTED** [review]

File: services/actions/notifier_helper.go (hunk 2 above)
Comment (denyskon) [reply services/actions/notifier_helper.go]: Oh, seems that I misunderstood their docs. I added pull_request_sync

File: services/actions/notifier_helper.go [hunk 3]
```diff
@@ -144,6 +146,22 @@ func notify(ctx context.Context, input *notifyInput) error {
...
+	if slices.Contains(skipRunEvents, input.Event) {
+		// skip runs with a configured skip-ci string in commit message if the event is push or pull_request(_sync)
+		// https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs
+		for _, s := range setting.Actions.SkipRunStrings {
+			if strings.Contains(commit.CommitMessage, s) {
+				log.Debug("repo %s with commit %s: skipped run because of %s string", input.Repo.RepoPath(), commit.ID, s)
+				return nil
+			}
+		}
+	}
```
Comment (delvh) [inline services/actions/notifier_helper.go]: I recommend extracting this into a function called `skipWorkflowsForCommit` returning a boolean.

**Review by delvh — APPROVED** [review]

**Review by nanguanlin6 — APPROVED** [review]
same as delvh but not a blocker.
BTW, probably should add more other tests in `action_trigger_test.go`. But not this PR's duty.

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -1396,6 +1396,7 @@ PROXY_HOSTS = *.github.com
 - `ZOMBIE_TASK_TIMEOUT`: **10m**: Timeout to stop the task which have running status, but haven't been updated for a long time
 - `ENDLESS_TASK_TIMEOUT`: **3h**: Timeout to stop the tasks which have running status and continuous updates, but don't end for a long time
 - `ABANDONED_JOB_TIMEOUT`: **24h**: Timeout to cancel the jobs which have waiting status, but haven't been picked by a runner for a long time
+- `SKIP_WORKFLOW_STRINGS`: **[skip ci]**, **[ci skip]**, **[no ci]**, **[skip actions]**, **[actions skip]**: Strings committers can place inside a commit message to skip executing the corresponding actions workflow
```
Comment (wxiaoguang) [inline docs/content/administration/config-cheat-sheet.en-us.md]: It doesn't seem right format for config value.  The "bold" part is the value, so IMO it should be:

```
**[skip ci],[ci skip],[no ci],[skip actions],[actions skip]**
```

File: services/actions/notifier_helper.go [hunk 5]
```diff
@@ -144,6 +146,10 @@ func notify(ctx context.Context, input *notifyInput) error {
 		return fmt.Errorf("gitRepo.GetCommit: %w", err)
 	}
 
+	if skipped := skipWorkflowsForCommit(input, commit); skipped {
```
Comment (wxiaoguang) [inline services/actions/notifier_helper.go]:
```suggestion
	if skipWorkflowsForCommit(input, commit) {
```

Why a variable .....

**Comment by wxiaoguang** [comment]
`app.example.ini` should also be updated.

**Comment by denyskon** [comment]
@wxiaoguang Done

File: custom/conf/app.example.ini [hunk 6]
```diff
@@ -2583,6 +2583,8 @@ LEVEL = Info
 ;ENDLESS_TASK_TIMEOUT = 3h
 ;; Timeout to cancel the jobs which have waiting status, but haven't been picked by a runner for a long time
 ;ABANDONED_JOB_TIMEOUT = 24h
+;; Strings committers can place inside a commit message to skip executing the corresponding actions workflow
+;SKIP_WORKFLOW_STRINGS = "[skip ci]", "[ci skip]", "[no ci]", "[skip actions]", "[actions skip]"
```
Comment (wxiaoguang) [inline custom/conf/app.example.ini]: That's not INI syntax .... IMO it should be:


```
;SKIP_WORKFLOW_STRINGS = "[skip ci],[ci skip],[no ci],[skip actions],[actions skip]"

or simply:

;SKIP_WORKFLOW_STRINGS = [skip ci],[ci skip],[no ci],[skip actions],[actions skip]
```

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 7]
```diff
@@ -1396,7 +1396,7 @@ PROXY_HOSTS = *.github.com
 - `ZOMBIE_TASK_TIMEOUT`: **10m**: Timeout to stop the task which have running status, but haven't been updated for a long time
 - `ENDLESS_TASK_TIMEOUT`: **3h**: Timeout to stop the tasks which have running status and continuous updates, but don't end for a long time
 - `ABANDONED_JOB_TIMEOUT`: **24h**: Timeout to cancel the jobs which have waiting status, but haven't been picked by a runner for a long time
-- `SKIP_WORKFLOW_STRINGS`: **[skip ci]**, **[ci skip]**, **[no ci]**, **[skip actions]**, **[actions skip]**: Strings committers can place inside a commit message to skip executing the corresponding actions workflow
+- `SKIP_WORKFLOW_STRINGS`: **"[skip ci]", "[ci skip]", "[no ci]", "[skip actions]", "[actions skip]"**: Strings committers can place inside a commit message to skip executing the corresponding actions workflow
```
Comment (wxiaoguang) [inline docs/content/administration/config-cheat-sheet.en-us.md]: See comment above.

File: custom/conf/app.example.ini (hunk 6 above)
Comment (denyskon) [reply custom/conf/app.example.ini]: I guess I'll never be a friend of ini 😆 Corrected

Comment (wxiaoguang) [reply custom/conf/app.example.ini]: It's still not fixed.

**Comment by wxiaoguang** [comment]
Still need to fix the config example

https://github.com/go-gitea/gitea/pull/28075#pullrequestreview-1737639131

File: custom/conf/app.example.ini (hunk 6 above)
Comment (denyskon) [reply custom/conf/app.example.ini]: Oops, forgot to push....

**Pull request merged**